src/modules/producer.js

```diff
@@ -0,0 +1,150 @@
+  if (rpcQueue.queue) return Promise.resolve(queue);
+
+  //we create the callback queue using base queue name + appending config hostname and :res for clarity
+  //ie. if hostname is gateway-http and queue is service-oauth, response queue will be service-oauth:gateway-http:res
+  //it is important to have different hostname or no hostname on each module sending message or there will be conflicts
+  var resQueue = queue + ':' + this.conn.config.hostname + ':res';
+  rpcQueue.queue = this.conn.get().then((channel) => {
+    return channel.assertQueue(resQueue, { durable: true, exclusive: true })
+      .then((_queue) => {
+        rpcQueue.queue = _queue.queue;
+
+        //if channel is closed, we want to make sure we cleanup the queue so future calls will recreate it
+        this.conn.addListener('close', () => { delete rpcQueue.queue; createRpcQueue.call(this, queue); });
+
+        return channel.consume(_queue.queue, maybeAnswer.call(this, queue), { noAck: true });
+      })
+      .then(() => {
+        return queue;
+      });
+    })
+    .catch(() => {
+      delete rpcQueue.queue;
+      return utils.timeoutPromise(this.conn.config.timeout).then(() => {
+        return createRpcQueue.call(this, queue);
+      });
+    });
+
+  return rpcQueue.queue;
+}
+
+/**
+ * Get a function to execute on channel consumer incoming message is received
+ * @param  {string} queue name of the queue where messages are SENT
+ * @return {function}       function executed by an amqp.node channel consume callback method
+ */
+function maybeAnswer(queue) {
+  var rpcQueue = amqpRPCQueues[queue];
+
+  return (msg) => {
+    //check the correlation ID sent by the initial message using RPC
+    var corrIdA = msg.properties.correlationId;
+    //if we found one, we execute the callback and delete it because it will never be received again anyway
+    Promise.resolve(parsers.in(msg))
+      .then((_msg) => {
+        this.conn.config.transport.info('amqp:producer', '[' + queue + '] < ', _msg);
+        return _msg;
+      })
+      .then(rpcQueue[corrIdA].resolve)
+      .then(() => {
+        delete rpcQueue[corrIdA];
+      })
+      .catch(this.conn.config.transport.error);
+  };
+}
+
+function publishOrSendToQueue(queue, msg, options) {
+  if (!options.routingKey) {
+    return this.channel.sendToQueue(queue, msg, options);
+  } else {
+    return this.channel.publish(queue, options.routingKey, msg, options);
+  }
+}
+
+/**
+ * Send message with or without rpc protocol, and check if RPC queues are created
+ * @param  {string} queue   the queue to send `msg` on
+ * @param  {any} msg     string, object, number.. anything bufferable/serializable
+ * @param  {object} options contain rpc property (if true, enable rpc for this message)
+ * @return {Promise}         Resolves when message is correctly sent, or when response is received when rpc is enabled
+ */
+function checkRpc (queue, msg, options) {
+  //messages are persistent
+  options.persistent = true;
```

Comment:
Well, yeah, something like this `const newOptions = Object.assign({},  options, {persistent: true});`